src/Hive.Utilities/LazyList.cs

```diff
@@ -0,0 +1,164 @@
+                    return;
+
+                srcEnum ??= source.GetEnumerator();
+
+                var result = true;
+                while ((cacheList.Count <= index || index == -1) && (result = srcEnum.MoveNext()))
+                {
+                    cacheList.Add(srcEnum.Current);
+                }
+
+                if (!result)
+                {
+                    // we finished evaluating the enumerator, so lets dispose it eagerly
+                    Interlocked.Exchange<IEnumerator<T>?>(ref srcEnum, null)?.Dispose();
+                    completed = true;
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public T this[int index]
+        {
+            get
+            {
+                if (index >= 0)
+                    ReadToIndex(index);
+                return cacheList[index];
+            }
+        }
+
+        /// <summary>
+        /// Attempts to get the count of items in the collection without
+        /// fully enumerating the input sequence.
+        /// </summary>
+        /// <value><see langword="null"/> if the length of the collection cannot be
+        /// determined without enumeration, otherwise the length of the collection.</value>
+        public int? NoEnumerateCount
+        {
+            get
+            {
+                if (source is IReadOnlyCollection<T> collection)
+                    return collection.Count;
+
+                if (completed)
+                    return cacheList.Count;
+
+                return null;
+            }
+        }
+
+        /// <inheritdoc/>
+        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+        public int Count
+        {
+            get
+            {
+                if (source is IReadOnlyCollection<T> collection)
+                    return collection.Count;
+
+                ReadToIndex(-1); // we have to evaluate the whole thing to get the count
+                return cacheList.Count;
+            }
+        }
+
+        /// <summary>
+        /// Gets an enumerator for this list.
+        /// </summary>
+        /// <returns>An enumerator for this list.</returns>
+        public Enumerator GetEnumerator() => new(this);
+        IEnumerator<T> IEnumerable<T>.GetEnumerator() => GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        /// <summary>
+        /// The enumerator type for <see cref="LazyList{T}"/>.
+        /// </summary>
+        public sealed class Enumerator : IEnumerator<T>
+        {
+            private readonly LazyList<T> list;
+            private int index = -1;
+
+            internal Enumerator(LazyList<T> list)
+                => this.list = list;
+
+            /// <inheritdoc/>
+            public T Current => list[index];
+            object? IEnumerator.Current => Current;
+
+            /// <inheritdoc/>
+            public bool MoveNext()
+            {
+                index++;
+                list.ReadToIndex(index);
```

Comment:
Not calling `ReadToIndex` means I need to duplicate most of the logic, which doesn't make sense. I can't imagine this would be too heavy of a bottleneck though.

I wrote it with this structure because it most clearly shows the order of operations. I'm not opposed to changing it, but I don't believe that there would be a benefit.